packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -579,84 +575,9 @@ function runDisconnectedCallback(vm: VM) {
     }
 }
 
-function runChildNodesDisconnectedCallback(vm: VM) {
-    const { velements: vCustomElementCollection } = vm;
-
-    // Reporting disconnection for every child in inverse order since they are
-    // inserted in reserved order.
-    for (let i = vCustomElementCollection.length - 1; i >= 0; i -= 1) {
-        const { elm } = vCustomElementCollection[i];
-
-        // There are two cases where the element could be undefined:
-        // * when there is an error during the construction phase, and an error
-        //   boundary picks it, there is a possibility that the VCustomElement
-        //   is not properly initialized, and therefore is should be ignored.
-        // * when slotted custom element is not used by the element where it is
-        //   slotted into it, as  a result, the custom element was never
-        //   initialized.
-        if (!isUndefined(elm)) {
-            const childVM = getAssociatedVMIfPresent(elm);
-
-            // The VM associated with the element might be associated undefined
-            // in the case where the VM failed in the middle of its creation,
-            // eg: constructor throwing before invoking super().
-            if (!isUndefined(childVM)) {
-                resetComponentStateWhenRemoved(childVM);
-            }
-        }
-    }
-}
-
-function runLightChildNodesDisconnectedCallback(vm: VM) {
-    const { aChildren: adoptedChildren } = vm;
-    recursivelyDisconnectChildren(adoptedChildren);
-}
-
-/**
- * The recursion doesn't need to be a complete traversal of the vnode graph,
- * instead it can be partial, when a custom element vnode is found, we don't
- * need to continue into its children because by attempting to disconnect the
- * custom element itself will trigger the removal of anything slotted or anything
- * defined on its shadow.
- */
-function recursivelyDisconnectChildren(vnodes: VNodes) {
-    for (let i = 0, len = vnodes.length; i < len; i += 1) {
-        const vnode = vnodes[i];
-
-        if (!isNull(vnode) && !isUndefined(vnode.elm)) {
-            switch (vnode.type) {
-                case VNodeType.Element:
-                    recursivelyDisconnectChildren(vnode.children);
-                    break;
-
-                case VNodeType.CustomElement: {
-                    const vm = getAssociatedVM(vnode.elm);
-                    resetComponentStateWhenRemoved(vm);
-                    break;
-                }
-            }
-        }
-    }
-}
-
-// This is a super optimized mechanism to remove the content of the root node (shadow root
-// for shadow DOM components and the root element itself for light DOM) without having to go
-// into snabbdom. Especially useful when the reset is a consequence of an error, in which case the
-// children VNodes might not be representing the current state of the DOM.
```

Comment:
Give this call out here, how is this PR addressing the fact that `children VNodes might not be representing the current state of the DOM.`?

Reply:
`unmountChildren` is effectively doing the same thing as what `resetComponentRoot` used to do. `unmountChildren` removes from the DOM all the direct children nodes attached to render root and also invoke descendant nodes `disconnectedCallback`s. 

The key difference here is the order of invocation. Previously `resetComponentRoot` removed all the render root children nodes and **then** invoked `disconnectedCallback` on all the descendants. With this PR, `unmountChildren` iterates over each children node, removes it, and invokes the `disconnectedCallback` on the child node descendant.  

It isn't clear to me how the current approach addresses the fact that `children VNodes might not be representing the current state of the DOM.`. I think this comment made sense when we were still using `[Element|ShadowRoot].proptotype.innerHTML` to clean up the render root content ([PR](https://github.com/salesforce/lwc/pull/1888/files#diff-03becdb9bf4c81b67cda99232c2db21cb07383df8500518d4f29ee687613dabbR518-R526)), but it's not the case anymore.